Avoid unnecessary imports for HLG Layer unpacking and materialization

rjzamora opened the pull request:
Motivated by #7374

Adds `layer_materialize_module` and `layer_materialize_class` properties to `highlevelgraph.Layer`.  These properties are used within `HighLevelGraph.__dask_distributed_pack__` so that a different class (usually a light-weight parent class) can be used for unpacking and graph materialization.

In order to leverage the new properties for shuffle operations, this PR adds a new `dask/layers.py` file, where lightweight versions of `SimpleShuffleLayer` and `ShuffleLayer` have been defined.  The "full" versions of these classes, in `dask.dataframe.shuffle`,  are now subclasses.  The new lightweight classes only include the required logic for unpacking and layer materialization, and they only import functions/classes from `dask.dask` and `tlz`.

Since we cannot import functions like `_concat` and `shuffle_group` in `layers.py`, we also use `run_ext_function` (new in this PR) for all materialized tasks so that we can build the graph with serialized functions.  **Note**: If we try to specify a serialized function as a callable object in a task, the `stringify_collection_keys` pass will misinterpret those tasks as collection keys.  I am very open to suggestions on other ways to avoid this problem.  However, if we do avoid the stringify issue, we would still need to modify `distributed.worker.dumps_task` to optionally skip the `dumps_func` step (since the functions will already need to be serialized).

rjzamora commented:
cc @jrbourbeau @ian-r-rose

ian-r-rose commented:
Thanks for the quick look at this @rjzamora. To me, the extra cognitive overhead of having two different flavors of, e.g., `SimpleShuffleLayer` is kind of unfortunate. Especially having a `layer_materialize_class`, for just one more level of indirection. If I understand correctly, you have done this to be able to pass in pickled versions of things like `_concat` from places that are actually allowed to import them.

I wonder if it would be possible to instead refer to those functions as string tuple of `(module, function)`. We've already crossed that bridge when it comes to packing/unacking the layers, what's the harm of one more :) Then your unpickling in `run_ext_function` would be replaced by something like an import/getattr.  I haven't really thought this through, however, and would be happy to be talked out of it.

rjzamora commented:
> To me, the extra cognitive overhead of having two different flavors of, e.g., SimpleShuffleLayer is kind of unfortunate. Especially having a layer_materialize_class, for just one more level of indirection. If I understand correctly, you have done this to be able to pass in pickled versions of things like _concat from places that are actually allowed to import them.

I think you have this mostly right for the particular case of shuffle.  We are moving the materialization component of `SimpleShuffleLayer` and `ShuffleLayer` into `layers.py`, because we need to have those methods defined in a class/file with minimal import requirements.  We are leaving the rest of the logic in the original `dataframe.shuffle.SimpleShuffleLayer` and `dataframe.shuffle.ShuffleLayer`.  For this specific case, the initialization, packing and culling logic could probably tweaked a bit to allow the entire class to be moved to `layers.py`, the `dask.dataframe.shuffle` function would just need to add a bit more logic to avoid the avoid doing things like passing in the pandas metadata.

I can understand the resistance to effectively splitting the class definition between two files.  My primary motivation was to keep as much dataframe/shuffle-specific logic in the `dataframe.shuffle` module itself.  It seems to me that we cannot use a `dask.dataframe.layers` module without importing the requirements for the `dask.dataframe` module.  So, if we want to avoid splitting class definitions, like this PR is currently doing, it seems that we may want to introduce a dedicated `highlevelgraph` directory structure to organize the large amount of HLG Layer/materialization logic that we eventually want to live in lightweight modules.

ian-r-rose commented on dask/layers.py:
```diff
@@ -0,0 +1,194 @@
+from .highlevelgraph import Layer
+
+
+def run_ext_function(func, *args):
+    if isinstance(func, dict):
+        func = getattr(import_module(func["__module__"]), func["__name__"])
+    return func(*args)
```
I personally prefer this variety of the function over the pickle version, but curious to know what others think!

ian-r-rose commented on dask/dataframe/shuffle.py:
```diff
@@ -134,33 +134,14 @@ def __dask_distributed_pack__(
             "name_input": self.name_input,
             "meta_input": to_serialize(self.meta_input),
             "parts_out": list(self.parts_out),
+            "concat_func": {"__module__": _concat.__module__, "__name__": "_concat"},
+            "getitem_func": {"__module__": "operator", "__name__": "getitem"},
+            "shuffle_group_func": {
+                "__module__": shuffle_group.__module__,
+                "__name__": "shuffle_group",
+            },
```
Should we have a name for this? Indirect import? Quoted import? It's starting to feel like a new flavor of task, perhaps worth formalizing in some way?

It is *almost* not necessary to have this  in `dask/dataframe/shuffle` at this point, no?

rjzamora replied:
> Should we have a name for this? Indirect import? Quoted import? It's starting to feel like a new flavor of task, perhaps worth formalizing in some way?

Right - Once we nail down what we actually want this to look like, I agree that we should clearly document and possibly formalize this approach in some way.

> It is almost not necessary to have this in dask/dataframe/shuffle at this point, no?

Right.  I am still open minded to the idea of moving everything into `layers.py`.  However, I personally feel that there are a variety of reasons not to do this.  In general, I feel that the path forward is likely to be much easier if we only require the unpacking and materialization logic to live in the lightweight (`layers.py`) module.  For shuffle, the only stickling point is the empty pd/cudf.DataFrame metadata that we need within `_construct_graph`.  When the `__dask_distributed_pack__` method is defined within the `dask.dataframe` module, it is easy to simply serialize this object.  However, if that definition was forced to live in `layers.py`, we wouldn't really be able to use the robust DataFrame serialization machinery.  I think we could certainly brainstorm ways to get around this, but I suspect that similar challenges will keep popping up.

ian-r-rose replied:
Ooh, good point. That actually opens a lot of doors :)

rjzamora replied:
so, to clarify, we can indeed move the shuffle layers entirely into `layers.py`.  Should we go ahead and do this?  That file may eventually become quite bloated, but it shouldn;t be muich work to cross that bridge later.

rjzamora replied:
I went ahead and moved "everything" over to `layers.py` (this includes `ShuffleLayer`, `SimpleShuffleLayer`, `BlockwiseCreateArray`, and `BroadcastJoinLayer`).  I think you are right about keeping the entire Layer definition in one place :)

jrbourbeau replied:
+1 thanks @rjzamora!

jrbourbeau reviewed (commented):
Could you add `dask.layers` to the set of modules on this line

https://github.com/dask/dask/blob/a1187b13321d69565b9c21359d739c239bd04c65/continuous_integration/scripts/test_imports.sh#L17

That will ensure we can import `dask.layers` with only Dask's required dependencies. This won't cover non top-level imports (e.g. importing something inside a `__dask_distributed_pack__` method) but it does provide us some good automated `import` coverage

rjzamora commented on dask/tests/test_distributed.py:
```diff
@@ -398,3 +335,93 @@ async def test_annotation_pack_unpack(c, s, a, b):
+    return da.ones((100,)) + da.zeros((100,))
+
+
+@pytest.mark.parametrize(
+    "op,lib",
+    [
+        (_dataframe_shuffle, "pandas."),
+        (_dataframe_broadcast_join, "pandas."),
+        (_array_creation, "numpy."),
+    ],
+)
+@pytest.mark.parametrize("optimize_graph", [True, False])
+def test_scheduler_highlevel_graph_unpack_import(op, lib, optimize_graph, loop):
+    # Test to highlight that in some cases we import modules like pandas
+    # on the scheduler when unpacking some HighLevelGraphs.
+    # This is a problem because pandas may not be installed in the process
+    # where the scheduler is running -- leading to an ImportError.
```
@jrbourbeau - I tried working in your testing idea from #7374 here.  Let me know if you would like me to focus on testing in this PR, or if you would like the changes to go in that PR with `xfail` (and then we can simply remove the `xfail` in this PR)

jrbourbeau replied:
Including this here sounds great -- thanks

jrbourbeau commented on dask/tests/test_distributed.py:
```diff
@@ -398,3 +335,93 @@ async def test_annotation_pack_unpack(c, s, a, b):
+        for mod in set(sys.modules):
+            if not mod.startswith(self.pattern):
+                self.start_modules.add(mod)
+            else:
+                # Maually remove the target library
+                sys.modules.pop(mod)
```
Modifying `sys.modules` in this way makes me a little uncomfortable -- though maybe this is actually fine

Since the `cluster` fixture creates a new process with the launched scheduler I wouldn't expect anything that's not a `distributed` dependency to be imported. What happens if we instead just `assert` that none of the initial `sys.modules` start with `self.pattern`?

rjzamora replied:
Sorry - I should have commented about this earlier.  I did start off by adding the assertion you suggest above.  However, the test was failing for the `_array_creation` case (pandas is not a problem).  It seems that numpy is imported somewhere outside of this test, and so I added the logic to manually "clean" `sys.modules` (rather than looking into it further) :/

jrbourbeau commented on dask/layers.py:
```diff
@@ -0,0 +1,837 @@
+#
+
+
+class CallableLazyImport:
+    """Function Wrapper for Lazy Importing"""
+
+    def __init__(self, function_path):
+        split_path = function_path.split(".")
+        self.module = ".".join(split_path[:-1])
+        self.name = split_path[-1]
+
+    def __call__(self, *args, **kwargs):
+        func = getattr(import_module(self.module), self.name)
+        return func(*args, **kwargs)
```
This looks new. Just to confirm, this is so we can materialize the graph on the scheduler without importing, say, `pandas`?

rjzamora replied:
Right - I am very interested in feedback on this. The idea is to have a "callable" at graph materialization time without requiring the scheduler to acutally import anything from a something like pandas/numpy/dask.dataframe/dask.array.

rjzamora replied:
Hmm - I actually think we can do that with a small tweak (by adding a deserializing parameter, as we do in Blockwise). I’ll try this out tonight.

jrbourbeau commented on dask/tests/test_distributed.py:
```diff
@@ -398,3 +335,93 @@ async def test_annotation_pack_unpack(c, s, a, b):
     assert annotations == {"workers": {"n": ("alice",)}}
+
+
+class SchedulerImportCheck(SchedulerPlugin):
```
What do you think about moving this to a new `test_layers.py` module?

rjzamora replied:
Good call - done.

jrbourbeau commented on dask/layers.py:
```diff
@@ -0,0 +1,837 @@
+        shape,
+        chunks,
+    ):
+        # self.name = name
```
I think this actually predates this PR

```suggestion
```

jrbourbeau reviewed (commented):
Thanks for taking this on @rjzamora! Generally I like the approach of moving layers to a separate module where we can test top-level imports.

Because of the large `diff` would you mind giving a brief high-level summary of the changes? It looks like the `Layer` classes were moved around and we introduced `CallableLazyImport` -- was there anything else that changed?

rjzamora commented on dask/tests/test_distributed.py:
```diff
@@ -117,6 +118,7 @@ def test_fused_blockwise_dataframe_merge(c, fuse):
 
 
 def test_futures_to_delayed_bag(c):
+    db = pytest.importorskip("dask.bag")
```
Oops, this line must have been added accidentally when I checked out your changes.

rjzamora commented on dask/tests/test_distributed.py:
```diff
@@ -398,3 +335,93 @@ async def test_annotation_pack_unpack(c, s, a, b):
+    async def start(self, scheduler):
+        # Record the modules that have been imported when the scheduler starts
+        self.start_modules = set()
+        for mod in set(sys.modules):
```
Doesn't seem to work without the `set` (or with `mod in sys.modules.keys()`).  My understanding is that, since we are modifying `sys.modules`,  we may need the `set` op to make a copy?

jrbourbeau replied:
Ah, my bad -- I didn't think about that

rjzamora commented:
>Because of the large diff would you mind giving a brief high-level summary of the changes? It looks like the Layer classes were moved around and we introduced CallableLazyImport -- was there anything else that changed?

Thanks for the review @jrbourbeau !

Yes - The high-level overview is that this PR simply moves existing Layer logic into the new `layers.py` file. For example:

- `dask.dataframe.shuffle.SimpleShuffleLayer` -> `dask.layers.SimpleShuffleLayer`
- `dask.dataframe.shuffle.ShuffleLayer` -> `dask.layers.ShuffleLayer`
- `dask.dataframe.multi.BroadcastJoinLayer` -> `dask.layers.BroadcastJoinLayer`
- `dask.array.blockwise.CreateArrayDeps` -> `dask.layers.CreateArrayDeps`
- `dask.array.blockwise.BlockwiseCreateArray` -> `dask.layers.BlockwiseCreateArray`

I tried to avoid changing the code, when possible.  However, in order to avoid importing functions like `dask.dataframe.shuffle.shuffle_group` during graph materialization, I introduced a `CallableLazyImport` class.  This purpose of this class is to delay the need for any specialized functions to be imported (from modules that may not be supported on the scheduler) until the task is executed on the worker.

This PR also uses your testing idea in a new `test_layers.py` module.

jrbourbeau commented:
Also cc @ian-r-rose if you get a chance to look at this

ian-r-rose commented on dask/layers.py:
```diff
@@ -0,0 +1,833 @@
+            {io_name: self.nchunks},
+            io_deps={
+                io_name: (
+                    "dask.layers",
+                    "CreateArrayDeps",
```
I'm reminded of the syntax for entry points, e.g. `dask.layers:CreateArrayDeps`. I've never particularly liked the `:`, but perhaps there is a reason for it that I am unaware of.

rjzamora replied:
Yeah - I agree that it is worth the extra `split` logic to use the fully-qualified names.  However, I agree it would be less confusing to make that change in a separate PR.

rjzamora commented on dask/layers.py:
```diff
@@ -0,0 +1,833 @@
+        # )
+
+        # if state["annotations"]:
+        #     cls.unpack_annotations(annotations, state["annotations"], raw.keys())
```
I think there are a few things going on here, and I'm not sure if the current logic is "wrong."  This code was originally copy-pasteed (mostly) from the blockwise and/or shuffle code.  It seems that recent Layer/annotation work modified the blockwise and shuffle `__dask_distributed_unpack__` definitions, but not the broadcast join (not sure whether or not the broadcast-join PR was merged after those changes).

rjzamora commented on dask/layers.py:
```diff
@@ -0,0 +1,833 @@
+    def __init__(self, function_path):
+        split_path = function_path.split(".")
+        self.module = ".".join(split_path[:-1])
+        self.name = split_path[-1]
```
Good call - This seems to be what `import_term` does (which is what we now use here).

rjzamora commented:
Okay - the last batch of changes avoids the use of `CallableLazyImport` unless we know the graph is being materialized on the scheduler.  Also, when that wrapper **is** used, we now take advantage of `distributed.utils.import_term`.

jrbourbeau commented on dask/layers.py:
```diff
@@ -0,0 +1,849 @@
+        state.update(merge_kwargs)
+
+        # Materialize the layer
+        raw = cls(**state).construct_graph(deserializing=True)
```
```suggestion
        raw = cls(**state)._construct_graph(deserializing=True)
```

jrbourbeau reviewed (approved):
Thanks for all your work on this @rjzamora!
